twake/frontend/src/app/scenes/Apps/Messages/Message/ActivityMessage.tsx

```diff
@@ -0,0 +1,264 @@
+            <Typography.Text strong className="small-x-margin">
+              {connector[0].name}
+            </Typography.Text>,
+            <Typography.Text strong>{resource.name}</Typography.Text>,
+          ],
+        );
+      }
+
+      if (activity.context.type === 'remove') {
+        return translateUsingReactNode(
+          'scenes.apps.messages.message.activity_message.a_deleted_channel_tab',
+          [
+            <span style={{ marginRight: 8, lineHeight: 0 }}>
+              {generateTypographyName(activity.actor.id)}
+            </span>,
+            <Typography.Text strong className="small-x-margin">
+              {connector[0].name}
+            </Typography.Text>,
+            <Typography.Text strong>{resource.name}</Typography.Text>,
+          ],
+        );
+      }
+    }
+  };
+
+  const channelConnectorCreatedOrDeleted = (activity: ActivityType) => {
+    if (activity.context.array) {
+      const resource = activity.context.array[0].resource as ChannelType;
+      // WorkspaceApps.getApp() doesn't work
+      const connector = WorkspacesApps.getApps().filter(app =>
+        resource.connectors?.includes(app.id),
+      );
+
+      if (connector.length) {
+        if (activity.context.type === 'add') {
+          return translateUsingReactNode(
+            'scenes.apps.messages.message.activity_message.a_created_channel_connector',
+            [
+              <span style={{ marginRight: 8, lineHeight: 0 }}>
+                {generateTypographyName(activity.actor.id)}
+              </span>,
+              <Typography.Text strong>{connector[0].name}</Typography.Text>,
+            ],
+          );
+        }
+
+        if (activity.context.type === 'remove') {
+          return translateUsingReactNode(
+            'scenes.apps.messages.message.activity_message.a_deleted_channel_connector',
+            [
+              <span style={{ marginRight: 8, lineHeight: 0 }}>
+                {generateTypographyName(activity.actor.id)}
+              </span>,
+              <Typography.Text strong>{connector[0].name}</Typography.Text>,
+            ],
+          );
+        }
+      }
+    }
+  };
+
+  const compute = () => {
+    switch (props.activity?.type) {
+      case ChannelActivityEnum.CHANNEL_MEMBER_CREATED:
+        return memberJoinedOrInvited(props.activity);
+      case ChannelActivityEnum.CHANNEL_MEMBER_DELETED:
+        return memberLeftOrRemoved(props.activity);
+      case ChannelActivityEnum.CHANNEL_UPDATED:
+        return channelNameOrDescription(props.activity);
+      case ChannelActivityEnum.CHANNEL_TAB_CREATED:
+      case ChannelActivityEnum.CHANNEL_TAB_DELETED:
+        return channelTabCreatedOrDeleted(props.activity);
+      case ChannelActivityEnum.CHANNEL_CONNECTOR_CREATED:
+      case ChannelActivityEnum.CHANNEL_CONNECTOR_DELETED:
+        return channelConnectorCreatedOrDeleted(props.activity);
+      default:
+        return 'Channel Activity not found';
+    }
+  };
```

Comment:
I am not a big fan of switch. You can let it like this but a good way to achieve this is to use a Map like:

```
const process = new Map<string, function>();
process.set(ChannelActivityEnum.CHANNEL_MEMBER_CREATED, memberJoinedOrInvited);
// put all

// then when you need to process:

return process.get(props.activity?.type)?(props.activity);
```

Reply:
Updated, thank you !